Userland/Libraries/LibWeb/Page/EventHandler.cpp

```diff
@@ -1086,6 +1096,18 @@ EventResult EventHandler::handle_keydown(UIEvents::KeyCode key, u32 modifiers, u
             return EventResult::Handled;
         }
 
+        if (!should_ignore_keydown_event(code_point, modifiers) && node.is_editable()) {
+            EventResult before_input_event_result;
+            if (key == UIEvents::KeyCode::Key_Return) {
+                before_input_event_result = fire_input_event(UIEvents::EventNames::beforeinput, UIEvents::InputTypes::insertParagraph, m_navigable, code_point);
+            } else {
+                before_input_event_result = fire_input_event(UIEvents::EventNames::beforeinput, UIEvents::InputTypes::insertText, m_navigable, code_point);
+            }
+            if (before_input_event_result == EventResult::Cancelled) {
+                return EventResult::Cancelled;
+            }
+        }
```

Comment:
Feels fragile having this here, if another key handler is inserted after the Key_Return handler below. Let's just bundle the `insertParagraph` event within the branch that handles return, and the `insertText` within the branch that handles text?

Reply:
As an aside, this method is getting a bit hard to follow...as a future cleanup, we might want to add a TRY-like macro that fires events and returns automatically if cancelled, so all of these are simple one-liners

(I'm envisioning something like):
```c++
FIRE(input_event(UIEvents::EventNames::beforeinput, UIEvents::InputTypes::insertParagraph, m_navigable, code_point));
```

Reply:
good idea, I added a FIRE macro.